### PR TITLE
[clang][x86] Add constexpr support for ADC/SBB + ADX intrinsics

### PR DESCRIPTION
**Pull request by RKSimon**
ADC and ADX use the same internal intrinsics - for testing I've taken the same approach as the generic builtin overflow tests, putting the intrinsics in a constexpr test wrapper and comparing the carry/result value pair.

I've added the addcarry/subborrow intrinsics to the clang language extension list - I'm not sure if we want to add all ISA intrinsics to the list (although we can if people think it useful?), but I felt we should at least include the baseline x86 intrinsics.

### CODE REVIEW

File: clang/lib/AST/ExprConstant.cpp [hunk 1]
```diff
@@ -13464,6 +13464,54 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
...
+                     (RHS.zext(BitWidth + 1) + (CarryIn.ugt(0) ? 1 : 0));
+
+    APInt Result = ExResult.extractBits(BitWidth, 0);
+    uint64_t CarryOut = ExResult.extractBitsAsZExtValue(1, BitWidth);
+
+    APValue APV{APSInt(Result, /*isUnsigned=*/true)};
+    if (!handleAssignment(Info, E, ResultLValue, ResultType, APV))
+      return false;
+    return Success(CarryOut, E);
+  }
```
Comment (goldsteinn) [inline clang/lib/AST/ExprConstant.cpp:13513]: Seems like a lot of dup code between these two cases. How about just using a bool to decide on add/sub between `LHS` and `RHS + Carry`?

File: clang/lib/AST/ExprConstant.cpp [hunk 2]
```diff
@@ -13464,6 +13464,38 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
...
+            : (LHS.zext(BitWidth + 1) - (RHS.zext(BitWidth + 1) + CarryInBit));
+
+    APInt Result = ExResult.extractBits(BitWidth, 0);
+    uint64_t CarryOut = ExResult.extractBitsAsZExtValue(1, BitWidth);
+
+    APValue APV{APSInt(Result, /*isUnsigned=*/true)};
+    if (!handleAssignment(Info, E, ResultLValue, ResultType, APV))
+      return false;
+    return Success(CarryOut, E);
+  }
```
Comment (goldsteinn) [inline clang/lib/AST/ExprConstant.cpp:13497]: LG although this isn't really a part of the code base I'm qualified to sign off on

**Comment by RKSimon** [comment]
@phoebewang Are you OK with this?

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**